docs/usage/config.md
# Configuration

## Synopsis

```text
$HOME/.config/syncthing
$HOME/Library/Application Support/Syncthing
%AppData%/Syncthing
%localappdata%/Syncthing
```

## Description

Syncthing uses a single directory to store configuration, crypto keys and index caches. The location defaults to `$HOME/.config/syncthing` \(Unix-like\), `$HOME/Library/Application Support/Syncthing` \(Mac\), `%AppData%/Syncthing` \(Windows XP\) or `%LocalAppData%/Syncthing` \(Windows 7+\). It can be changed at runtime using the `-home` flag. In this directory the following files are located:

* config.xml: The configuration file, in XML format.
* cert.pem, key.pem: The device's RSA public and private key. These form the basis for the device ID. The key must be kept private.
* https-cert.pem, https-key.pem: The certificate and key for HTTPS GUI connections. These may be replaced with a custom certificate for HTTPS as desired.
* index-\*.db: A directory holding the database with metadata and hashes of the files currently on disk and available from peers.
* csrftokens.txt: A list of recently issued CSRF tokens \(for protection against browser cross site request forgery\).

## Config File Format

The following shows an example of the default configuration file \(IDs will differ\):

\`\`\` {.sourceCode .xml}

 basic 1 0 0 0 random false 0 0 -1 false false false 25 .stfolder dynamic false 127.0.0.1:8384 k1dnz1Dd0rzTBjjFFh7CXPnrF12C49B1 default default default true true 21027 \[ff12::8384\]:21027 0 0 60 true 10 true true 60 30 10 0 0 LFWe2vn3 [https://data.syncthing.net/newdata](https://data.syncthing.net/newdata) false 1800 true 12 false 24 false 5 false 1 [https://upgrades.syncthing.net/meta.json](https://upgrades.syncthing.net/meta.json) false 10 0 auto default 24 false 25 false true 128 128 ~ 0 true

```text
Configuration Element
---------------------

``` {.sourceCode .xml}
<configuration version="26">
    <folder></folder>
    <device></device>
    <gui></gui>
    <options></options>
    <ignoredDevice>5SYI2FS-LW6YAXI-JJDYETS-NDBBPIO-256MWBO-XDPXWVG-24QPUM4-PDW4UQU</ignoredDevice>
    <ignoredFolder>bd7q3-zskm5</ignoredDevice>
</configuration>
```

This is the root element. It has one attribute:

version

: The config version. Increments whenever a change is made that requires migration from previous formats.

It contains the elements described in the following sections and these two additional child elements:

ignoredDevice

: Contains the ID of the device that should be ignored. Connection attempts from this device are logged to the console but never displayed in the web GUI.

ignoredFolder

: Contains the ID of the folder that should be ignored. This folder will always be skipped when advertised from a remote device, i.e. this will be logged, but there will be no dialog about it in the web GUI.

## Folder Element

\`\`\` {.sourceCode .xml}

 basic 1 0 0 0 random false 0 0 -1 false false false 25 .stfolder

```text
One or more `folder` elements must be present in the file. Each element
describes one folder. The following attributes may be set on the
`folder` element:

id

:   The folder ID, must be unique. (mandatory)

label

:   The label of a folder is a human readable and descriptive local
    name. May be different on each device, empty, and/or identical to
    other folder labels. (optional)

path

:   The path to the directory where the folder is stored on this device;
    not sent to other devices. (mandatory)

type

:   Controls how the folder is handled by Syncthing. Possible values
    are:

    sendreceive

    :   The folder is in default mode. Sending local and accepting
        remote changes. Note that this type was previously called
        \"readwrite\" which is deprecated but still accepted in incoming
        configs.

    sendonly

    :   The folder is in \"send only\" mode \-- it will not be modified
        by Syncthing on this device. Note that this type was previously
        called \"readonly\" which is deprecated but still accepted in
        incoming configs.

    receiveonly

    :   The folder is in \"receive only\" mode \-- it will not propagate
        changes to other devices.

rescanIntervalS

:   The rescan interval, in seconds. Can be set to zero to disable when
    external plugins are used to trigger rescans.

fsWatcherEnabled

:   If enabled this detects changes to files in the folder and scans
    them.

::: {#fsWatcherDelayS}

fsWatcherDelayS

:   The duration during which changes detected are accumulated, before a
    scan is scheduled (only takes effect if `fsWatcherEnabled` is true).

ignorePerms

:   True if the folder should ignore permissions.

autoNormalize

:   Automatically correct UTF-8 normalization errors found in file
    names.
:::

The following child elements may exist:

device

:   These must have the `id` attribute and can have an `introducedBy`
    attribute, identifying the device that introduced us to share this
    folder with the given device. If the original introducer unshares
    this folder with this device, our device will follow and unshare the
    folder (subject to skipIntroductionRemovals being false on the
    introducer device). All mentioned devices are those that will be
    sharing the folder in question. Each mentioned device must have a
    separate `device` element later in the file. It is customary that
    the local device ID is included in all folders. Syncthing will
    currently add this automatically if it is not present in the
    configuration file.

minDiskFree

:   The minimum required free space that should be available on the disk
    this folder resides. The folder will be stopped when the value drops
    below the threshold. Accepted units are `%`, `kB`, `MB`, `GB` and
    `TB`. Set to zero to disable.

versioning

:   Specifies a versioning configuration.

::: {.seealso}
`versioning`{.interpreted-text role="ref"}
:::

copiers, pullers, hashers

:   The number of copier, puller and hasher routines to use, or zero for
    the system determined optimum. These are low level performance
    options for advanced users only; do not change unless requested to
    or you\'ve actually read and understood the code yourself. :)

order

:   The order in which needed files should be pulled from the cluster.
    The possibles values are:

    random

    :   Pull files in random order. This optimizes for balancing
        resources among the devices in a cluster.

    alphabetic

    :   Pull files ordered by file name alphabetically.

    smallestFirst, largestFirst

    :   Pull files ordered by file size; smallest and largest first
        respectively.

    oldestFirst, newestFirst

    :   Pull files ordered by modification time; oldest and newest first
        respectively.

    Note that the scanned files are sent in batches and the sorting is
    applied only to the already discovered files. This means the sync
    might start with a 1 GB file even if there is 1 KB file available on
    the source device until the 1 KB becomes known to the pulling
    device.

ignoreDelete

:   When set to true, this device will pretend not to see instructions
    to delete files from other devices.

scanProgressIntervalS

:   The interval with which scan progress information is sent to the
    GUI. Zero means the default value (two seconds).

pullerPauseS

:   Tweak for rate limiting the puller when it retries pulling files.
    Don\'t change these unless you know what you\'re doing.

maxConflicts

:   The maximum number of conflict copies to keep around for any given
    file. The default, -1, means an unlimited number. Setting this to
    zero disables conflict copies altogether.

disableSparseFiles

:   By default, blocks containing all zeroes are not written, causing
    files to be sparse on filesystems that support the concept. When set
    to true, sparse files will not be created.

disableTempIndexes

:   By default, devices exchange information about blocks available in
    transfers that are still in progress. When set to true, such
    information is not exchanged for this folder.

paused

:   True if this folder is (temporarily) suspended.

weakHashThresholdPct

:   Use weak hash if more than the given percentage of the file has
    changed. Set to -1 to always use weak hash. Default value is 25.

markerName

:   Name of a directory or file in the folder root to be used as
    `marker-faq`{.interpreted-text role="ref"}. Default is
    \".stfolder\".

fsync

:   ::: {.deprecated}
    v0.14.37
    :::

    Transfer updated (from other devices) files to permanent storage
    before committing the changes to the internal database.

pullerSleepS

:   ::: {.deprecated}
    v0.14.41
    :::

    Tweak for rate limiting the puller. Don\'t change these unless you
    know what you\'re doing.

Device Element
--------------

``` {.sourceCode .xml}
<device id="5SYI2FS-LW6YAXI-JJDYETS-NDBBPIO-256MWBO-XDPXWVG-24QPUM4-PDW4UQU" name="syno" compression="metadata" introducer="false" skipIntroductionRemovals="false" introducedBy="2CYF2WQ-AKZO2QZ-JAKWLYD-AGHMQUM-BGXUOIS-GYILW34-HJG3DUK-LRRYQAR">
    <address>dynamic</address>
</device>
<device id="2CYF2WQ-AKZO2QZ-JAKWLYD-AGHMQUM-BGXUOIS-GYILW34-HJG3DUK-LRRYQAR" name="syno local" compression="metadata" introducer="false">
    <address>tcp://192.0.2.1:22001</address>
    <paused>true<paused>
    <allowedNetwork>192.168.0.0/16<allowedNetwork>
</device>
```

One or more `device` elements must be present in the file. Each element describes a device participating in the cluster. It is customary to include a `device` element for the local device; Syncthing will currently add one if it is not present. The following attributes may be set on the `device` element:

id

: The device ID. This must be written in canonical form, that is without any spaces or dashes. \(mandatory\)

name

: A friendly name for the device. \(optional\)

compression

: Whether to use protocol compression when sending messages to this device. The possible values are:

```text
metadata

:   Compress metadata packets, such as index information. Metadata
    is usually very compression friendly so this is a good default.

always

:   Compress all packets, including file data. This is recommended
    if the folders contents are mainly compressible data such as
    documents or text files.

never

:   Disable all compression.
```

introducer

: Set to true if this device should be trusted as an introducer, i.e. we should copy their list of devices per folder when connecting.

::: {.seealso} `introducer`{.interpreted-text role="ref"} :::

skipIntroductionRemovals

: Set to true if you wish to follow only introductions and not de-introductions. For example, if this is set, we would not remove a device that we were introduced to even if the original introducer is no longer listing the remote device as known.

introducedBy

: Defines which device has introduced us to this device. Used only for following de-introductions.

certName

: The device certificate common name, if it is not the default \"syncthing\".

From following child elements at least one `address` child must exist.

address

: Contains an address or host name to use when attempting to connect to this device. Entries other than `dynamic` must be prefixed with `tcp://` \(dual-stack\), `tcp4://` \(IPv4 only\) or `tcp6://` \(IPv6 only\). Note that IP addresses need not use tcp4/tcp6; these are optional. Accepted formats are:

```text
IPv4 address (`tcp://192.0.2.42`)

:   The default port (22000) is used.

IPv4 address and port (`tcp://192.0.2.42:12345`)

:   The address and port is used as given.

IPv6 address (`tcp://[2001:db8::23:42]`)

:   The default port (22000) is used. The address must be enclosed
    in square brackets.

IPv6 address and port (`tcp://[2001:db8::23:42]:12345`)

:   The address and port is used as given. The address must be
    enclosed in square brackets.

Host name (`tcp6://fileserver`)

:   The host name will be used on the default port (22000) and
    connections will be attempted only via IPv6.

Host name and port (`tcp://fileserver:12345`)

:   The host name will be used on the given port and connections
    will be attempted via both IPv4 and IPv6, depending on name
    resolution.

`dynamic`

:   The word `dynamic` (without `tcp://` prefix) means to use local
    and global discovery to find the device.

You can set multiple addresses *and* combine it with the `dynamic`
keyword for example:

``` {.sourceCode .xml}
<device id="...">
    <address>tcp://192.0.2.1:22001</address>
    <address>tcp://192.0.1.254:22000</address>
    <address>dynamic</address>
</device>
```
```

paused

: True if synchronization with this devices is \(temporarily\) suspended.

allowedNetwork

: If given, this restricts connections to this device to only this network \(see `allowed-networks`{.interpreted-text role="ref"}\).

## GUI Element

\`\`\` {.sourceCode .xml}

 127.0.0.1:8384 l7jSbCqPD95JYZ0g8vi4ZLAMg3ulnN1b default

```text
There must be exactly one `gui` element. The GUI configuration is also
used by the `rest-api`{.interpreted-text role="ref"} and the
`event-api`{.interpreted-text role="ref"}. The following attributes may
be set on the `gui` element:

enabled

:   If not `true`, the GUI and API will not be started.

tls

:   If set to `true`, TLS (HTTPS) will be enforced. Non-HTTPS requests
    will be redirected to HTTPS. When this is set to `false`, TLS
    connections are still possible but it is not mandatory.

debugging

:   This enables `profiling`{.interpreted-text role="ref"} and
    additional debugging endpoints in the `rest-api`{.interpreted-text
    role="ref"}.

The following child elements may be present:

address

:   Set the listen address. One address element must be present. Allowed
    address formats are:

    IPv4 address and port (`127.0.0.1:8384`)

    :   The address and port is used as given.

    IPv6 address and port (`[::1]:8384`)

    :   The address and port is used as given. The address must be
        enclosed in square brackets.

    Wildcard and port (`0.0.0.0:12345`, `[::]:12345`, `:12345`)

    :   These are equivalent and will result in Syncthing listening on
        all interfaces via both IPv4 and IPv6.

    UNIX socket location (`/var/run/st.sock`)

    :   If the address is an absolute path it is interpreted as the path
        to a UNIX socket. (Added in v0.14.52.)

user

:   Set to require authentication.

password

:   Contains the bcrypt hash of the real password.

apikey

:   If set, this is the API key that enables usage of the REST
    interface.

insecureAdminAccess

:   If true, this allows access to the web GUI from outside (i.e. not
    localhost) without authorization. A warning will displayed about
    this setting on startup.

theme

:   The name of the theme to use.

authMode

:   Authentication mode to use. If not present authentication mode
    (static) is controlled by presence of user/passward fields for
    backward compatibility.

    static

    :   Authentication using user and password.

    ldap

    :   LDAP authentication. Requires ldap top level config section to
        be present.

LDAP Element
------------

``` {.sourceCode .xml}
<ldap>
    <address>localhost:389</address>
    <bindDN>cn=%s,ou=users,dc=syncthing,dc=net</bindDN>
    <transport>nontls</transport>
    <insecureSkipVerify>false</insecureSkipVerify> 
</ldap>
```

The `ldap` element contains LDAP configuration options.

address

: LDAP server address \(server:port\).

bindDN

: BindDN for user authentication. Special %s variable shoild be used to pass username to LDAP.

transport

> nontls
>
> : Non secure connection.
>
> tls
>
> : TLS secured connection.
>
> starttls
>
> : StartTLS connection mode.
>
> insecureSkipVerify

: Skip verification \(true or false\).

## Options Element

\`\`\` {.sourceCode .xml}

 default default true true 21027 \[ff12::8384\]:21027 0 0 60 true 10 true true 60 30 10 0 [https://data.syncthing.net/newdata](https://data.syncthing.net/newdata) false 1800 true 12 24 false 5 false 1 [https://api.github.com/repos/syncthing/syncthing/releases?per\_page=30](https://api.github.com/repos/syncthing/syncthing/releases?per_page=30) false 10 ~

\`\`\`

The `options` element contains all other global configuration options.

listenAddress

: The listen address for incoming sync connections. See `listen-addresses`{.interpreted-text role="ref"} for allowed syntax.

globalAnnounceServer

: A URI to a global announce \(discovery\) server, or the word `default` to include the default servers. Any number of globalAnnounceServer elements may be present. The syntax for non-default entries is that of a HTTP or HTTPS URL. A number of options may be added as query options to the URL: `insecure` to prevent certificate validation \(required for HTTP URLs\) and `id=<device ID>` to perform certificate pinning. The device ID to use is printed by the discovery server on startup.

globalAnnounceEnabled

: Whether to announce this device to the global announce \(discovery\) server, and also use it to look up other devices.

localAnnounceEnabled

: Whether to send announcements to the local LAN, also use such announcements to find other devices.

localAnnouncePort

: The port on which to listen and send IPv4 broadcast announcements to.

localAnnounceMCAddr

: The group address and port to join and send IPv6 multicast announcements on.

maxSendKbps

: Outgoing data rate limit, in kibibytes per second.

maxRecvKbps

: Incoming data rate limits, in kibibytes per second.

reconnectionIntervalS

: The number of seconds to wait between each attempt to connect to currently unconnected devices.

relaysEnabled

: When true, relays will be connected to and potentially used for device to device connections.

relayReconnectIntervalM

: Sets the interval, in minutes, between relay reconnect attempts.

startBrowser

: Whether to attempt to start a browser to show the GUI when Syncthing starts.

natEnabled

: Whether to attempt to perform a UPnP and NAT-PMP port mapping for incoming sync connections.

natLeaseMinutes

: Request a lease for this many minutes; zero to request a permanent lease.

natRenewalMinutes

: Attempt to renew the lease after this many minutes.

natTimeoutSeconds

: When scanning for UPnP devices, wait this long for responses.

urAccepted

: Whether the user has accepted to submit anonymous usage data. The default, `0`, mean the user has not made a choice, and Syncthing will ask at some point in the future. `-1` means no, a number above zero means that that version of usage reporting has been accepted.

urSeen

: The highest usage reporting version that has already been shown in the web GUI.

urUniqueID


lib/config/config.go
// Copyright (C) 2014 The Syncthing Authors.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at http://mozilla.org/MPL/2.0/.

// Package config implements reading and writing of the syncthing configuration file.
package config

import (
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io"
	"io/ioutil"
	"net/url"
	"os"
	"path"
	"sort"
	"strings"

	"github.com/syncthing/syncthing/lib/protocol"
	"github.com/syncthing/syncthing/lib/rand"
	"github.com/syncthing/syncthing/lib/util"
)

const (
	OldestHandledVersion = 10
	CurrentVersion       = 16
	MaxRescanIntervalS   = 365 * 24 * 60 * 60
)

var (
	// DefaultListenAddresses should be substituted when the configuration
	// contains <listenAddress>default</listenAddress>. This is done by the
	// "consumer" of the configuration as we don't want these saved to the
	// config.
	DefaultListenAddresses = []string{
		"tcp://0.0.0.0:22000",
		"dynamic+https://relays.syncthing.net/endpoint",
	}
	// DefaultDiscoveryServersV4 should be substituted when the configuration
	// contains <globalAnnounceServer>default-v4</globalAnnounceServer>.
	DefaultDiscoveryServersV4 = []string{
		"https://discovery-v4-2.syncthing.net/v2/?id=DVU36WY-H3LVZHW-E6LLFRE-YAFN5EL-HILWRYP-OC2M47J-Z4PE62Y-ADIBDQC", // 45.55.230.38, USA
		"https://discovery-v4-3.syncthing.net/v2/?id=VK6HNJ3-VVMM66S-HRVWSCR-IXEHL2H-U4AQ4MW-UCPQBWX-J2L2UBK-NVZRDQZ", // 128.199.95.124, Singapore
		"https://discovery-v4-4.syncthing.net/v2/?id=LYXKCHX-VI3NYZR-ALCJBHF-WMZYSPK-QG6QJA3-MPFYMSO-U56GTUK-NA2MIAW", // 95.85.19.244, NL
	}
	// DefaultDiscoveryServersV6 should be substituted when the configuration
	// contains <globalAnnounceServer>default-v6</globalAnnounceServer>.
	DefaultDiscoveryServersV6 = []string{
		"https://discovery-v6-2.syncthing.net/v2/?id=DVU36WY-H3LVZHW-E6LLFRE-YAFN5EL-HILWRYP-OC2M47J-Z4PE62Y-ADIBDQC", // 2604:a880:800:10::182:a001, USA
		"https://discovery-v6-3.syncthing.net/v2/?id=VK6HNJ3-VVMM66S-HRVWSCR-IXEHL2H-U4AQ4MW-UCPQBWX-J2L2UBK-NVZRDQZ", // 2400:6180:0:d0::d9:d001, Singapore
		"https://discovery-v6-4.syncthing.net/v2/?id=LYXKCHX-VI3NYZR-ALCJBHF-WMZYSPK-QG6QJA3-MPFYMSO-U56GTUK-NA2MIAW", // 2a03:b0c0:0:1010::4ed:3001, NL
	}
	// DefaultDiscoveryServers should be substituted when the configuration
	// contains <globalAnnounceServer>default</globalAnnounceServer>.
	DefaultDiscoveryServers = append(DefaultDiscoveryServersV4, DefaultDiscoveryServersV6...)

	// DefaultTheme is the default and fallback theme for the web UI.
	DefaultTheme = "default"
)

func New(myID protocol.DeviceID) Configuration {
	var cfg Configuration
	cfg.Version = CurrentVersion
	cfg.OriginalVersion = CurrentVersion

	util.SetDefaults(&cfg)
	util.SetDefaults(&cfg.Options)
	util.SetDefaults(&cfg.GUI)

	// Can't happen.
	if err := cfg.prepare(myID); err != nil {
		panic("bug: error in preparing new folder: " + err.Error())
	}

	return cfg
}

func ReadXML(r io.Reader, myID protocol.DeviceID) (Configuration, error) {
	var cfg Configuration

	if err := xml.NewDecoder(r).Decode(&cfg); err != nil {
		return Configuration{}, err
	}
	cfg.OriginalVersion = cfg.Version

	if err := cfg.prepare(myID); err != nil {
		return Configuration{}, err
	}
	return cfg, nil
}

func ReadJSON(r io.Reader, myID protocol.DeviceID) (Configuration, error) {
	var cfg Configuration

	bs, err := ioutil.ReadAll(r)
	if err != nil {
		return Configuration{}, err
	}

	if err := json.Unmarshal(bs, &cfg); err != nil {
		return Configuration{}, err
	}
	cfg.OriginalVersion = cfg.Version

	if err := cfg.prepare(myID); err != nil {
		return Configuration{}, err
	}
	return cfg, nil
}

type Configuration struct {
	Version        int                   `xml:"version,attr" json:"version"`
	Folders        []FolderConfiguration `xml:"folder" json:"folders"`
	Devices        []DeviceConfiguration `xml:"device" json:"devices"`
	GUI            GUIConfiguration      `xml:"gui" json:"gui"`
	Options        OptionsConfiguration  `xml:"options" json:"options"`
	IgnoredDevices []protocol.DeviceID   `xml:"ignoredDevice" json:"ignoredDevices"`
	XMLName        xml.Name              `xml:"configuration" json:"-"`

	OriginalVersion int `xml:"-" json:"-"` // The version we read from disk, before any conversion
}

func (cfg Configuration) Copy() Configuration {
	newCfg := cfg

	// Deep copy FolderConfigurations
	newCfg.Folders = make([]FolderConfiguration, len(cfg.Folders))
	for i := range newCfg.Folders {
		newCfg.Folders[i] = cfg.Folders[i].Copy()
	}

	// Deep copy DeviceConfigurations
	newCfg.Devices = make([]DeviceConfiguration, len(cfg.Devices))
	for i := range newCfg.Devices {
		newCfg.Devices[i] = cfg.Devices[i].Copy()
	}

	newCfg.Options = cfg.Options.Copy()

	// DeviceIDs are values
	newCfg.IgnoredDevices = make([]protocol.DeviceID, len(cfg.IgnoredDevices))
	copy(newCfg.IgnoredDevices, cfg.IgnoredDevices)

	return newCfg
}

func (cfg *Configuration) UnmarshalJSON(data []byte) error {
	type methodless Configuration
	util.SetDefaults(cfg)
	return json.Unmarshal(data, (*methodless)(cfg))
}

func (cfg *Configuration) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	type methodless Configuration
	util.SetDefaults(cfg)
	return d.DecodeElement((*methodless)(cfg), &start)
}

func (cfg *Configuration) WriteXML(w io.Writer) error {
	e := xml.NewEncoder(w)
	e.Indent("", "    ")
	err := e.Encode(cfg)
	if err != nil {
		return err
	}
	_, err = w.Write([]byte("\n"))
	return err
}

func (cfg *Configuration) prepare(myID protocol.DeviceID) error {
	var myName string

	// Ensure this device is present in the config
	for _, device := range cfg.Devices {
		if device.DeviceID == myID {
			goto found
		}
	}

	myName, _ = os.Hostname()
	cfg.Devices = append(cfg.Devices, DeviceConfiguration{
		DeviceID: myID,
		Name:     myName,
	})

found:

	if err := cfg.clean(); err != nil {
		return err
	}

	// Ensure that we are part of the devices
	for i := range cfg.Folders {
		cfg.Folders[i].Devices = ensureDevicePresent(cfg.Folders[i].Devices, myID)
	}

	return nil
}

func (cfg *Configuration) clean() error {
	util.FillNilSlices(&cfg.Options)

	// Initialize any empty slices
	if cfg.Folders == nil {
		cfg.Folders = []FolderConfiguration{}
	}
	if cfg.IgnoredDevices == nil {
		cfg.IgnoredDevices = []protocol.DeviceID{}
	}
	if cfg.Options.AlwaysLocalNets == nil {
		cfg.Options.AlwaysLocalNets = []string{}
	}
	if cfg.Options.UnackedNotificationIDs == nil {
		cfg.Options.UnackedNotificationIDs = []string{}
	}

	// Prepare folders and check for duplicates. Duplicates are bad and
	// dangerous, can't currently be resolved in the GUI, and shouldn't
	// happen when configured by the GUI. We return with an error in that
	// situation.
	seenFolders := make(map[string]struct{})
	for i := range cfg.Folders {
		folder := &cfg.Folders[i]
		folder.prepare()

		if _, ok := seenFolders[folder.ID]; ok {
			return fmt.Errorf("duplicate folder ID %q in configuration", folder.ID)
		}
		seenFolders[folder.ID] = struct{}{}
	}

	cfg.Options.ListenAddresses = util.UniqueStrings(cfg.Options.ListenAddresses)
	cfg.Options.GlobalAnnServers = util.UniqueStrings(cfg.Options.GlobalAnnServers)

	if cfg.Version > 0 && cfg.Version < OldestHandledVersion {
		l.Warnf("Configuration version %d is deprecated. Attempting best effort conversion, but please verify manually.", cfg.Version)
	}

	// Upgrade configuration versions as appropriate
	if cfg.Version <= 10 {
		convertV10V11(cfg)
	}
	if cfg.Version == 11 {
		convertV11V12(cfg)
	}
	if cfg.Version == 12 {
		convertV12V13(cfg)
	}
	if cfg.Version == 13 {
		convertV13V14(cfg)
	}
	if cfg.Version == 14 {
		convertV14V15(cfg)
	}
	if cfg.Version == 15 {
		convertV15V16(cfg)
	}

	// Build a list of available devices
	existingDevices := make(map[protocol.DeviceID]bool)
	for _, device := range cfg.Devices {
		existingDevices[device.DeviceID] = true
	}

	// Ensure that the device list is free from duplicates
	cfg.Devices = ensureNoDuplicateDevices(cfg.Devices)

	sort.Sort(DeviceConfigurationList(cfg.Devices))
	// Ensure that any loose devices are not present in the wrong places
	// Ensure that there are no duplicate devices
	// Ensure that the versioning configuration parameter map is not nil
	for i := range cfg.Folders {
		cfg.Folders[i].Devices = ensureExistingDevices(cfg.Folders[i].Devices, existingDevices)
		cfg.Folders[i].Devices = ensureNoDuplicateFolderDevices(cfg.Folders[i].Devices)
		if cfg.Folders[i].Versioning.Params == nil {
			cfg.Folders[i].Versioning.Params = map[string]string{}
		}
		sort.Sort(FolderDeviceConfigurationList(cfg.Folders[i].Devices))
	}

	// An empty address list is equivalent to a single "dynamic" entry
	for i := range cfg.Devices {
		n := &cfg.Devices[i]
		if len(n.Addresses) == 0 || len(n.Addresses) == 1 && n.Addresses[0] == "" {
			n.Addresses = []string{"dynamic"}
		}
	}

	// Very short reconnection intervals are annoying
	if cfg.Options.ReconnectIntervalS < 5 {
		cfg.Options.ReconnectIntervalS = 5
	}

	if cfg.GUI.APIKey == "" {
		cfg.GUI.APIKey = rand.String(32)
	}

	// The list of ignored devices should not contain any devices that have
	// been manually added to the config.
	newIgnoredDevices := []protocol.DeviceID{}
	for _, dev := range cfg.IgnoredDevices {
		if !existingDevices[dev] {
			newIgnoredDevices = append(newIgnoredDevices, dev)
		}
	}
	cfg.IgnoredDevices = newIgnoredDevices

	return nil
}

func convertV14V15(cfg *Configuration) {
	// Undo v0.13.0 broken migration

	for i, addr := range cfg.Options.GlobalAnnServers {
		switch addr {
		case "default-v4v2/":
			cfg.Options.GlobalAnnServers[i] = "default-v4"
		case "default-v6v2/":
			cfg.Options.GlobalAnnServers[i] = "default-v6"
		}
	}

	cfg.Version = 15
}

func convertV15V16(cfg *Configuration) {
	// Triggers a database tweak
	cfg.Version = 16
}

func convertV13V14(cfg *Configuration) {
	// Not using the ignore cache is the new default. Disable it on existing
	// configurations.
	cfg.Options.CacheIgnoredFiles = false

	// Migrate UPnP -> NAT options
	cfg.Options.NATEnabled = cfg.Options.DeprecatedUPnPEnabled
	cfg.Options.DeprecatedUPnPEnabled = false
	cfg.Options.NATLeaseM = cfg.Options.DeprecatedUPnPLeaseM
	cfg.Options.DeprecatedUPnPLeaseM = 0
	cfg.Options.NATRenewalM = cfg.Options.DeprecatedUPnPRenewalM
	cfg.Options.DeprecatedUPnPRenewalM = 0
	cfg.Options.NATTimeoutS = cfg.Options.DeprecatedUPnPTimeoutS
	cfg.Options.DeprecatedUPnPTimeoutS = 0

	// Replace the default listen address "tcp://0.0.0.0:22000" with the
	// string "default", but only if we also have the default relay pool
	// among the relay servers as this is implied by the new "default"
	// entry.
	hasDefault := false
	for _, raddr := range cfg.Options.DeprecatedRelayServers {
		if raddr == "dynamic+https://relays.syncthing.net/endpoint" {
			for i, addr := range cfg.Options.ListenAddresses {
				if addr == "tcp://0.0.0.0:22000" {
					cfg.Options.ListenAddresses[i] = "default"
					hasDefault = true
					break
				}
			}
			break
		}
	}

	// Copy relay addresses into listen addresses.
	for _, addr := range cfg.Options.DeprecatedRelayServers {
		if hasDefault && addr == "dynamic+https://relays.syncthing.net/endpoint" {
			// Skip the default relay address if we already have the
			// "default" entry in the list.
			continue
		}
		if addr == "" {
			continue
		}
		cfg.Options.ListenAddresses = append(cfg.Options.ListenAddresses, addr)
	}

	cfg.Options.DeprecatedRelayServers = nil

	// For consistency
	sort.Strings(cfg.Options.ListenAddresses)

	var newAddrs []string
	for _, addr := range cfg.Options.GlobalAnnServers {
		uri, err := url.Parse(addr)
		if err != nil {
			// That's odd. Skip the broken address.
			continue
		}
		if uri.Scheme == "https" {
			uri.Path = path.Join(uri.Path, "v2") + "/"
			addr = uri.String()
		}

		newAddrs = append(newAddrs, addr)
	}
	cfg.Options.GlobalAnnServers = newAddrs

	for i, fcfg := range cfg.Folders {
		if fcfg.DeprecatedReadOnly {
			cfg.Folders[i].Type = FolderTypeReadOnly
		} else {
			cfg.Folders[i].Type = FolderTypeReadWrite
		}
		cfg.Folders[i].DeprecatedReadOnly = false
	}
	// v0.13-beta already had config version 13 but did not get the new URL
	if cfg.Options.ReleasesURL == "https://api.github.com/repos/syncthing/syncthing/releases?per_page=30" {
		cfg.Options.ReleasesURL = "https://upgrades.syncthing.net/meta.json"
	}

	cfg.Version = 14
}

func convertV12V13(cfg *Configuration) {
	if cfg.Options.ReleasesURL == "https://api.github.com/repos/syncthing/syncthing/releases?per_page=30" {
		cfg.Options.ReleasesURL = "https://upgrades.syncthing.net/meta.json"
	}

	cfg.Version = 13
}

func convertV11V12(cfg *Configuration) {
	// Change listen address schema
	for i, addr := range cfg.Options.ListenAddresses {
		if len(addr) > 0 && !strings.HasPrefix(addr, "tcp://") {
			cfg.Options.ListenAddresses[i] = util.Address("tcp", addr)
		}
	}

	for i, device := range cfg.Devices {
		for j, addr := range device.Addresses {
			if addr != "dynamic" && addr != "" {
				cfg.Devices[i].Addresses[j] = util.Address("tcp", addr)
			}
		}
	}

	// Use new discovery server
	var newDiscoServers []string
	var useDefault bool
	for _, addr := range cfg.Options.GlobalAnnServers {
		if addr == "udp4://announce.syncthing.net:22026" {
			useDefault = true
		} else if addr == "udp6://announce-v6.syncthing.net:22026" {
			useDefault = true
		} else {
			newDiscoServers = append(newDiscoServers, addr)
		}
	}
	if useDefault {
		newDiscoServers = append(newDiscoServers, "default")
	}
	cfg.Options.GlobalAnnServers = newDiscoServers

	// Use new multicast group
	if cfg.Options.LocalAnnMCAddr == "[ff32::5222]:21026" {
		cfg.Options.LocalAnnMCAddr = "[ff12::8384]:21027"
	}

	// Use new local discovery port
	if cfg.Options.LocalAnnPort == 21025 {
		cfg.Options.LocalAnnPort = 21027
	}

	// Set MaxConflicts to unlimited
	for i := range cfg.Folders {
		cfg.Folders[i].MaxConflicts = -1
	}

	cfg.Version = 12
}

func convertV10V11(cfg *Configuration) {
	// Set minimum disk free of existing folders to 1%
	for i := range cfg.Folders {
		cfg.Folders[i].MinDiskFreePct = 1
	}
	cfg.Version = 11
}

func ensureDevicePresent(devices []FolderDeviceConfiguration, myID protocol.DeviceID) []FolderDeviceConfiguration {
	for _, device := range devices {
		if device.DeviceID.Equals(myID) {
			return devices
		}
	}

	devices = append(devices, FolderDeviceConfiguration{
		DeviceID: myID,
	})

	return devices
}

func ensureExistingDevices(devices []FolderDeviceConfiguration, existingDevices map[protocol.DeviceID]bool) []FolderDeviceConfiguration {
	count := len(devices)
	i := 0
loop:
	for i < count {
		if _, ok := existingDevices[devices[i].DeviceID]; !ok {
			devices[i] = devices[count-1]
			count--
			continue loop
		}
		i++
	}
	return devices[0:count]
}

func ensureNoDuplicateFolderDevices(devices []FolderDeviceConfiguration) []FolderDeviceConfiguration {
	count := len(devices)
	i := 0
	seenDevices := make(map[protocol.DeviceID]bool)
loop:
	for i < count {
		id := devices[i].DeviceID
		if _, ok := seenDevices[id]; ok {
			devices[i] = devices[count-1]
			count--
			continue loop
		}
		seenDevices[id] = true
		i++
	}
	return devices[0:count]
}

func ensureNoDuplicateDevices(devices []DeviceConfiguration) []DeviceConfiguration {
	count := len(devices)
	i := 0
	seenDevices := make(map[protocol.DeviceID]bool)
loop:
	for i < count {
		id := devices[i].DeviceID
		if _, ok := seenDevices[id]; ok {
			devices[i] = devices[count-1]
			count--
			continue loop
		}
		seenDevices[id] = true
		i++
	}
	return devices[0:count]
}
